src/server/frontend_wayland/layer_shell_v1.cpp

```diff
@@ -160,11 +160,7 @@ void mf::LayerSurfaceV1::set_keyboard_interactivity(uint32_t keyboard_interactiv
 
 void mf::LayerSurfaceV1::get_popup(struct wl_resource* popup)
 {
-    auto popup_window_role =
-        static_cast<WindowWlSurfaceRole*>(
-            static_cast<XdgPopupStable*>(
-                static_cast<wayland::XdgPopup*>(
-                    wl_resource_get_user_data(popup))));
+    auto popup_window_role = static_cast<WindowWlSurfaceRole*>(XdgPopupStable::from(popup));
```

Comment:
Do we need `static_cast` to reach an accessible base class?

Reply:
I just did that so I could use `auto`.

Reply:
```C++
    auto* const popup_window_role = XdgPopupStable::from(popup);
```
or, if the exact type is really important:
```c++
    WindowWlSurfaceRole* const popup_window_role = XdgPopupStable::from(popup);
```